bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/common/AbstractExpressionPart.java

```diff
@@ -0,0 +1,180 @@
+
+    private String part;
+    private BoundedIntegerSet valueSet;
+
+    public AbstractExpressionPart(String s) throws ParseException {
+        this.set(s);
+        this.parse();
+    }
+
+    @Override
+    public void set(String s) {
+        this.setPart(s);
+    }
+
+    /**
+     * @return the set of values that are valid for the expression part
+     */
+    public BoundedIntegerSet getValueSet() {
+        return valueSet;
+    }
+
+    /**
+     * @param set the set of values that the epxression part should take into consideration
+     */
+    public void setValueSet(BoundedIntegerSet set) {
+        this.valueSet = set;
+    }
+
+    abstract BoundedIntegerSet initializeValueSet();
+
+    /**
+     * A <code>BoundedIntegerSet</code> is a set of Integers that is bound by a minimum and maximum value, that does
+     * allow or not allow negaive values, and that is either starts from index 0 or 1
+     *
+     * @author Karel Goderis
+     *
+     */
+    protected class BoundedIntegerSet extends TreeSet<Integer> {
+        private static final long serialVersionUID = 19296179649170335L;
+        protected final int max;
+        protected final int min;
+        protected final boolean negative;
+        protected boolean is1indexed;
+
+        BoundedIntegerSet(final int min, final int max, final boolean negativeValuesAllowed, final boolean is1indexed) {
+            this.min = min;
+            this.max = max;
+            this.negative = negativeValuesAllowed;
+            this.is1indexed = is1indexed;
+        }
+
+        /**
+         * Validates if an integer can be added to the set
+         *
+         * @param integer - the integer to validate
+         */
+        private void validate(final Integer integer) {
+            if (negative) {
+                final int abs = Math.abs(integer);
+                if (abs < min || abs > max) {
+                    throw new IllegalArgumentException("Invalid integer value (value not in range [" + (-max) + ", "
+                            + -min + "] U [" + min + ", " + max + "])");
+                }
+            } else {
+                if (integer < min || integer > max) {
+                    throw new IllegalArgumentException(
+                            "Invalid integer value (value not in range [" + min + ", " + max + "])");
+                }
+            }
+        }
+
+        @Override
+        public boolean add(final Integer e) {
+            validate(e);
+            return super.add(e);
+        }
+
+        /**
+         * Adds a series of integers to the set, starting from start, and ending at end, with a given increment
+         *
+         * @param start - value to start at
+         * @param end - value to end at (inclusive)
+         * @param increment - increment to advance
+         * @return true if the addition of integers was successful
+         */
+        public boolean add(final Integer start, final Integer end, final Integer increment) {
```

Comment:
Well, the general idea is that you parse a part of an expression, which should finally result in a series of dates that are included in the expression. So for example, for a day in the month, you would have a bound integer set with min = 1, max = 31. and thus an expression expressing something relative to a day in the month, could result in a set with the integers 5,6,7 to indicate that the 5th, 6th and the 7th day of the month are relevant for generating candidate dates. For months, you would accept integer values 1 to 12. and so forth. Some expressions work with relative time indications, e.g. -3 to indicate the third last of the given series (hence, negative values in the boundintegerset)